### PR TITLE
Full ROI Chain

### PR DESCRIPTION
**Pull request by bfonta**
This PR brings the full ROI chain with a new seeding:
- fully functional hexagonal seeding
- clustering on the x/z,y/z space taking ROI seeds and ROI TCs (or all TCs if preferred) as input
- resolution and distribution plotting utilities, including interactive visualization of hexagonal grids and events

- important fix for the "max_energy" clustering algorithm (see ```bye_splits/tasks/cluster.py```): the old version assigns all TCs (withing a maximum distance) to the same, most-energetic seed. The new version correctly prioritizes seeds based on energy, so that the seeds that are not the most energetic are still assigned some TCs.

The chain was tested with multiple parameters from start to finish.

Screenshot from 2023-04-09 23-24-26

Screenshot from 2023-04-09 23-26-38

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -48,7 +48,7 @@ varEvents:
     z: 'good_tc_z'
     phi: 'good_tc_phi'
     eta: 'good_tc_eta'
-    idcl: 'good_tc_multicluster_id'
+    idcl: 'good_tc_cluster_id'
```
Comment (isehle) [inline config.yaml:51]: Was this change done on purpose? We want the `multicluster_id`, yes?

File: config.yaml [hunk 2]
```diff
@@ -159,13 +161,19 @@ valid:
   FillOutComp: 'fill_comp'
   FillOut: 'fill'
 
-valid_roi:
+valid_cluster:
   ClusterOutForValidation: 'cluster_valid_roi'
```
Comment (isehle) [inline config.yaml:165]: Shouldn't we have one cluster validation that doesn't include the `_roi` tag, for the default chain?

**Review by isehle — COMMENTED** [review]
Overall it's looking really great! Passed through the ROI related files quite quickly, but the updates elsewhere look good and shouldn't affect anything on my side. One general question—you switched from the ```with <file> as <var>``` syntax to ```<var> = <file>``` in a few places; is this primarily for debugging purposes or is there a deeper reason?

File: config.yaml (hunk 1 above)
Comment (bfonta) [reply config.yaml:51]: You are right, I keep modifying it for local tests.

**Comment by bfonta** [comment]
@isehle No important reason; I did it simply to avoid the hard-to-read indentation when one has multiple ```open()``` calls in the same ```with``` clause. By the way, we now have to explicitly call ```.close()```.

File: config.yaml (hunk 2 above)
Comment (bfonta) [reply config.yaml:165]: [We do](https://github.com/bfonta/bye_splits/blob/feature/seeding/config.yaml#L159-L169). I changed the naming just to clarify to which part of the reco chain this validation corresponds; ```valid_roi``` was too generic, and could refer to the ROI finder, seeding or clustering in the ROI chain.

Comment (isehle) [reply config.yaml:165]: Ah yes, I missed the first one. For consistencies sake though, I'd suggest they both use the same name (either `ClusterOutValidation` or `ClusterOutForValidation`, but not both—unless there's a specific reason why not.)

**Comment by isehle** [comment]
@bfonta I see, yes that's helpful enough. And yup, I'm aware that `.close()` is necessary now, but thanks for the reminder!

File: config.yaml (hunk 2 above)
Comment (bfonta) [reply config.yaml:165]: Fair enough.

**Review by isehle — APPROVED** [review]
Looks great!

**Review by isehle — APPROVED** [review]

File: bye_splits/tasks/roi.py [hunk 3]
```diff
@@ -52,21 +51,21 @@ def roi_dummy_calculator(tcs, k=4, threshold=20, nearby=True):
...
+    roi_df = pd.DataFrame()
+    
     module_sums = create_module_sums(input_df)
     module_ROI = list(module_sums[module_sums.values >= threshold].index)
-
     if nearby:
-        selected_modules = []
-        for module in module_ROI:
-            nearby_modules = [(module[0]+i, module[1]+j) for i in [-1, 0, 1] for j in [-1, 0, 1] if i*j >= 0]
-            skim = (module_sums.index.isin(nearby_modules)) & (module_sums.values > 0.3 * module_sums[module]) & (module_sums.values > 10)
```
Comment (mchiusi) [inline bye_splits/tasks/roi.py:63]: Have you decided to pass every time all the six nearby modules _plus_ the central one?
Do you apply an energy selection?

**Review by mchiusi — COMMENTED** [review]
Overall it's looking great.

File: bye_splits/tasks/roi.py (hunk 3 above)
Comment (mchiusi) [reply bye_splits/tasks/roi.py:63]: Ok perfect. That's all I wanted to know.

**Review by mchiusi — APPROVED** [review]

**Pull request merged**